cpp/include/cudf/detail/gather.cuh

```diff
@@ -518,6 +506,58 @@ void gather_bitmask(table_view const& source,
   }
 }
 
+template <typename MapItRoot>
+struct column_gatherer_impl<struct_view, MapItRoot> {
+  std::unique_ptr<column> operator()(column_view const& column,
+                                     MapItRoot gather_map_begin,
+                                     MapItRoot gather_map_end,
+                                     bool nullify_out_of_bounds,
+                                     cudaStream_t stream,
+                                     rmm::mr::device_memory_resource* mr)
+  {
+    structs_column_view structs_column(column);
+    auto gather_map_size{std::distance(gather_map_begin, gather_map_end)};
+    if (gather_map_size == -1) {
+      return make_empty_column(
+        data_type{type_id::STRUCT});  // TODO: Construct empty member columns, if required.
+    }
+
+    std::vector<std::unique_ptr<cudf::column>> output_struct_members;
+    std::transform(structs_column.child_begin(),
+                   structs_column.child_end(),
+                   std::back_inserter(output_struct_members),
+                   [&gather_map_begin, &gather_map_end, nullify_out_of_bounds, stream, mr](
+                     cudf::column_view const& col) {
+                     return cudf::type_dispatcher(col.type(),
+                                                  column_gatherer{},
+                                                  col,
+                                                  gather_map_begin,
+                                                  gather_map_end,
+                                                  nullify_out_of_bounds,
+                                                  stream,
+                                                  mr);
+                   });
+
+    gather_bitmask(
```

Comment:
I'm not sure I've grokked this. Why would the bitmask be bashed?

Reply:
After the type dispatch call completes, the higher level gather call generates and overwrites the bitmask on the column that gets returned.  This is where it handles all the squirrely logic for nullify_out_of_bounds, etc.

https://github.com/rapidsai/cudf/blob/0d83d64d959efc96ec1e85437819c39b70643298/cpp/include/cudf/detail/gather.cuh#L570

Reply:
> the higher level gather call generates and overwrites the bitmask on the column that gets returned.

My apologies for not documenting this better.

`column_gatherer_impl<struct_view>::operator()` calls `gather_bitmask()` to set the null mask for its children, not itself. This code is very similar to how the null mask is set in `gather()`. The top level `struct` column's null mask is set to a default value (Line#555), with the assurance that this will be fixed up in `gather()` (Line#607).

Reply:
Ah, right.  I wasn't reading closely enough.